gradio/blocks.py

```diff
@@ -208,29 +208,46 @@ def render(self):
         self._id = Context.id
         Context.id += 1
 
-    def process_api(self, data: Dict[str, Any], username: str = None) -> Dict[str, Any]:
+    def process_api(
+        self,
+        data: Dict[str, Any],
+        username: str = None,
+        state: Optional[Dict[int, any]] = None,
+    ) -> Dict[str, Any]:
         raw_input = data["data"]
         fn_index = data["fn_index"]
+        session_hash = data["session_hash"]
         fn, preprocess, postprocess = self.fns[fn_index]
         dependency = self.dependencies[fn_index]
 
         if preprocess:
-            processed_input = [
-                self.blocks[input_id].preprocess(raw_input[i])
-                for i, input_id in enumerate(dependency["inputs"])
-            ]
+            processed_input = []
+            for i, input_id in enumerate(dependency["inputs"]):
+                block = self.blocks[input_id]
+                if getattr(block, "stateful", False):
+                    processed_input.append(state.get(input_id))
+                else:
+                    processed_input.append(block.preprocess(raw_input[i]))
             predictions = fn(*processed_input)
         else:
             predictions = fn(*raw_input)
         if len(dependency["outputs"]) == 1:
             predictions = (predictions,)
         if postprocess:
-            predictions = [
-                self.blocks[output_id].postprocess(predictions[i])
-                if predictions[i] is not None
-                else None
-                for i, output_id in enumerate(dependency["outputs"])
-            ]
+            output = []
+            for i, output_id in enumerate(dependency["outputs"]):
+                block = self.blocks[output_id]
+                if getattr(block, "stateful", False):
+                    state[output_id] = predictions[i]
+                    output.append(None)
+                else:
+                    output.append(
+                        block.postprocess(predictions[i])
+                        if predictions[i] is not None
+                        else None
+                    )
+            else:
+                output = predictions
```

Comment:
why do we override `output` after the loop is finished?

Reply:
Whoops typo - the return of this function should be output. Output is not overridden after the loop, its the 'else' clause where there is no postprocesssing, so the output is just the predictions.

Reply:
As far as I can see the else comes after the for loop and it enters to the `else` scope every time.
```
for loops also have an else clause which most of us are unfamiliar with. The else clause executes after the loop completes normally.
```